// @fsegurai/codemirror-theme-gruvbox-dark/dist/index.cjs
'use strict';

var view = require('@codemirror/view');
var language = require('@codemirror/language');
var highlight = require('@lezer/highlight');

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

/**
 * Enhanced Gruvbox Dark theme color definitions
 * --------------------------------------------
 * Colors organized by function with visual color blocks
 */
// Gruvbox base colors
const base00 = '#282828', // Background
base01 = '#3c3836', // Lighter background (popups, statuslines)
base02 = '#504945', // Selection background
base03 = '#665c54', // Comments, invisibles, line highlighting
base04 = '#7c6f64', // Dark foreground (status bars)
base05 = '#928374', // Comments, invisibles, line highlighting
// Light/foreground shades
base06 = '#fbf1c7', // Light foreground (preferbase08)
base07 = '#ebdbb2', // Light foreground (alternative)
// Accent colors
base08 = '#fb4934', // Keywords, storage, operator
base09 = '#b8bb26', // Strings, tag attributes
base0A = '#fabd2f', // Functions, tag names
base0B = '#83a598', // Variables
base0C = '#d3869b', // Numbers, special constants
base0D = '#8ec07c', // Types
base0E = '#fe8019'; // Cursor, constants
// UI specific colors
const invalid = base08, darkBackground = base01, highlightBackground = '#3c383660', // Line highlight with transparency
background = base00, tooltipBackground = base01, selection = base02, selectionMatch = '#665c5480', // Selection match background
cursor = base0E, // Cursor color
activeBracketBg = '#504945cc', // Active bracket background
activeBracketBorder = base0A, // Active bracket border
diagnosticWarning = base0A, // Warning color
linkColor = base0B, // Link color
visitedLinkColor = base0C; // Visited link color
// Diff/merge specific colors
const addedBackground = '#32361a80', // Dark base09 with transparency for insertions
removedBackground = '#3c1f1e80', // Dark base08 with transparency for deletions
addedText = '#b8bb26', // Bright base09 for added text
removedText = '#fb4934'; // Bright base08 for removed text
/**
 * Enhanced editor theme styles for Gruvbox Dark
 */
const gruvboxDarkTheme = view.EditorView.theme({
    // Base editor styles
    '&': {
        color: base07,
        backgroundColor: background,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: base00,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selection,
    },
    // Make sure selection appears above active line
    '.cm-selectionLayer': {
        zIndex: 100,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: '#b57614cc',
        outline: `1px solid ${base0A}`,
        color: base06,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: base06,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: base0E,
        color: base00,
        padding: generalSearchField.padding,
        '& span': {
            color: base00,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base07,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: tooltipBackground,
        color: base07,
        borderRadius: '4px',
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: `1px solid ${base03}`,
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: `1px solid ${base03}`,
    },
    '.cm-panel button': {
        backgroundColor: tooltipBackground,
        color: base07,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: base02,
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
        zIndex: 1,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: base01,
        color: base05,
        border: generalGutter.border,
        borderRight: `1px solid ${base02}`,
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: darkBackground,
        color: base07,
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base05,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base07,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}40`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}40`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: `1px solid ${base03}`,
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 2px 8px rgba(0, 0, 0, 0.3)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul': {
            backgroundColor: tooltipBackground,
            border: 'none',
        },
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: selection,
            color: base06,
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base05,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base05,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${invalid}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${diagnosticWarning}`,
        },
        '&-info': {
            borderLeft: `3px solid ${linkColor}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${invalid}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${diagnosticWarning}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: '#cc241d55',
        outline: `1px solid ${invalid}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: selectionMatch,
        outline: `1px solid ${base03}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: tooltipBackground,
        color: base05,
        border: `1px dotted ${base03}`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${base00}, 0 0 0 3px ${base0E}40`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: base01,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: base03,
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${base01}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: base04,
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: base05,
    },
}, { dark: true });
/**
 * Enhanced syntax highlighting for Gruvbox Dark theme
 */
const gruvboxDarkHighlightStyle = language.HighlightStyle.define([
    // Keywords and control flow
    { tag: highlight.tags.keyword, color: base08, fontWeight: 'bold' },
    { tag: highlight.tags.controlKeyword, color: base08, fontWeight: 'bold' },
    { tag: highlight.tags.moduleKeyword, color: base08, fontWeight: 'bold' },
    // Names and variables
    { tag: [highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName], color: base0B },
    { tag: [highlight.tags.variableName], color: base0B },
    { tag: [highlight.tags.propertyName], color: base0D, fontStyle: 'normal' },
    // Classes and types
    { tag: [highlight.tags.typeName], color: base0D },
    { tag: [highlight.tags.className], color: base0A, fontStyle: 'italic' },
    { tag: [highlight.tags.namespace], color: base0B, fontStyle: 'italic' },
    // Operators and punctuation
    { tag: [highlight.tags.operator, highlight.tags.operatorKeyword], color: base07 },
    { tag: [highlight.tags.bracket], color: base05 },
    { tag: [highlight.tags.brace], color: base05 },
    { tag: [highlight.tags.punctuation], color: base05 },
    // Functions and parameters
    { tag: [highlight.tags.function(highlight.tags.variableName), highlight.tags.labelName], color: base0A },
    { tag: [highlight.tags.definition(highlight.tags.variableName)], color: base0B },
    // Constants and literals
    { tag: highlight.tags.number, color: base0C },
    { tag: highlight.tags.changed, color: base0C },
    { tag: highlight.tags.annotation, color: invalid, fontStyle: 'italic' },
    { tag: highlight.tags.modifier, color: base0C, fontStyle: 'italic' },
    { tag: highlight.tags.self, color: base0C },
    { tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)], color: base0E },
    { tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)], color: base0E },
    // Strings and regex
    { tag: [highlight.tags.processingInstruction, highlight.tags.inserted], color: base09 },
    { tag: [highlight.tags.special(highlight.tags.string), highlight.tags.regexp], color: base09 },
    { tag: highlight.tags.string, color: base09 },
    // Punctuation and structure
    { tag: highlight.tags.definition(highlight.tags.typeName), color: base0D, fontWeight: 'bold' },
    // Comments and documentation
    { tag: highlight.tags.meta, color: base05 },
    { tag: highlight.tags.comment, fontStyle: 'italic', color: base05 },
    { tag: highlight.tags.docComment, fontStyle: 'italic', color: base05 },
    // HTML/XML elements
    { tag: [highlight.tags.tagName], color: base08 },
    { tag: [highlight.tags.attributeName], color: base0A },
    // Markdown and text formatting
    { tag: [highlight.tags.heading], fontWeight: 'bold', color: base0A },
    { tag: [highlight.tags.strong], fontWeight: 'bold', color: base0A },
    { tag: [highlight.tags.emphasis], fontStyle: 'italic', color: base09 },
    // Links and URLs
    {
        tag: [highlight.tags.link],
        color: visitedLinkColor,
        fontWeight: '500',
        textDecoration: 'underline',
        textUnderlinePosition: 'under',
    },
    {
        tag: [highlight.tags.url],
        color: linkColor,
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    {
        tag: [highlight.tags.invalid],
        color: base07,
        textDecoration: 'underline wavy',
        borderBottom: `1px wavy ${invalid}`,
    },
    { tag: [highlight.tags.strikethrough], color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: highlight.tags.constant(highlight.tags.name), color: base0E },
    { tag: highlight.tags.deleted, color: invalid },
    { tag: highlight.tags.squareBracket, color: base05 },
    { tag: highlight.tags.angleBracket, color: base05 },
    // Additional specific styles
    { tag: highlight.tags.monospace, color: base07 },
    { tag: [highlight.tags.contentSeparator], color: base0B },
    { tag: highlight.tags.quote, color: base05 },
]);
/**
 * Combined Gruvbox Dark theme extension
 */
const gruvboxDark = [
    gruvboxDarkTheme,
    language.syntaxHighlighting(gruvboxDarkHighlightStyle),
];
/**
 * Gruvbox Dark merge revert styles configuration
 */
const gruvboxDarkMergeStyles = {
    backgroundColor: tooltipBackground,
    borderColor: base03,
    buttonColor: base07,
    buttonHoverColor: base02,
};

exports.applyMergeRevertStyles = applyMergeRevertStyles;
exports.gruvboxDark = gruvboxDark;
exports.gruvboxDarkMergeStyles = gruvboxDarkMergeStyles;
