// @fsegurai/codemirror-theme-basic-dark/dist/index.cjs
'use strict';

var view = require('@codemirror/view');
var language = require('@codemirror/language');
var highlight = require('@lezer/highlight');

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

/**
 * Enhanced Basic Dark theme color palette
 * ------------------------------------------
 * Colors refined for better contrast and visual hierarchy
 */
// Base colors
const base00 = '#1e2124', // Background (slightly darker for better contrast)
base01 = '#e2e2e2', // Foreground (slightly brighter for better readability)
base02 = '#5c88da', // Selection elements (more vibrant blue)
base03 = '#b8b8b8', // Comments, line numbers (slightly lighter)
base04 = '#ffffff', // Cursor (pure white for better visibility)
base05 = '#e4e4e4', // Panels foreground
base06 = '#909090', // Comments (slightly more visible gray)
base07 = '#000000', // Pure black for contrast elements
base08 = '#e06c75', // Error, deleted (more vibrant red)
base09 = '#f39c12', // Number, boolean (warmer orange)
base0A = '#ffcb6b', // Keywords (warmer yellow)
base0B = '#98c379', // Strings (more vibrant green)
base0C = '#56b6c2', // Classes, types (cyan blue)
base0D = '#61afef', // Functions, methods (bright blue)
base0E = '#c678dd', // Operators, brackets (brighter purple)
base0F = '#be5046', // Special elements (darker red)
// UI-specific colors
invalid = '#e06c75', // Error highlighting (consistent red)
darkBackground = '#252529', // Panel background (slightly darker)
selectionBackground = '#3a5991aa', // Selection background (semi-transparent blue)
highlightBackground = '#3a3d4166', // Active line background (subtle blue-gray)
tooltipBackground = '#2a2c31', // Tooltip background (darker than the editor)
cursor = base04, activeBracketBg = '#3a599140', activeBracketBorder = base0E, 
// Diff/merge specific colors
addedBackground = '#1e462c50', // Dark green with transparency for insertions
removedBackground = '#5c1e2340', // Dark red with transparency for deletions
addedText = '#73c991', // Green for added text
removedText = '#f07178'; // Red for removed text
/**
 * Enhanced editor theme styles for Basic Dark
 */
const basicDarkTheme = view.EditorView.theme({
    // Base editor styles
    '&': {
        color: base01,
        backgroundColor: base00,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: base00,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selectionBackground,
        color: base01,
    },
    // Make sure selection appears above active line
    '.cm-selectionLayer': {
        zIndex: 100,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: '#4a74c480',
        outline: `1px solid ${base02}`,
        color: base01,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: base01,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: '#5c88da90',
        color: base01,
        padding: generalSearchField.padding,
        '& span': {
            color: base01,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base05,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: darkBackground,
        color: base05,
        borderRadius: '0 0 4px 4px',
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: '1px solid #3a3a3a',
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: '1px solid #3a3a3a',
    },
    '.cm-panel button': {
        backgroundColor: darkBackground,
        color: base05,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: '#35373d',
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
        zIndex: 1,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: darkBackground,
        color: base06,
        border: generalGutter.border,
        borderRight: '1px solid #35383d',
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: highlightBackground,
        color: base03,
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base06,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base01,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}40`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}40`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: '1px solid #3a3a3a',
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 2px 8px rgba(0, 0, 0, 0.3)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul': {
            backgroundColor: tooltipBackground,
            border: 'none',
        },
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: selectionBackground,
            color: base01,
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base06,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base06,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${invalid}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${base0A}`,
        },
        '&-info': {
            borderLeft: `3px solid ${base0D}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${invalid}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${base0A}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}60`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: '#e06c7540',
        outline: `1px solid ${invalid}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: '#4a74c440',
        outline: `1px solid ${base02}50`,
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: 'transparent',
        color: base02,
        border: `1px dotted ${base02}70`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${base00}, 0 0 0 3px ${base02}40`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: darkBackground,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: '#444448',
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${darkBackground}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: '#55555a',
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: '#b8b8b8',
    },
}, { dark: true });
/**
 * Enhanced syntax highlighting for Basic Dark theme
 */
const basicDarkHighlightStyle = language.HighlightStyle.define([
    // Keywords and control flow
    { tag: highlight.tags.keyword, color: base0A, fontWeight: 'bold' },
    { tag: highlight.tags.controlKeyword, color: base0A, fontWeight: 'bold' },
    { tag: highlight.tags.moduleKeyword, color: base0A, fontWeight: 'bold' },
    // Names and variables
    { tag: [highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName], color: base0D },
    { tag: [highlight.tags.variableName], color: base0D },
    { tag: [highlight.tags.propertyName], color: base0C, fontStyle: 'normal' },
    // Classes and types
    { tag: [highlight.tags.typeName], color: base0C },
    { tag: [highlight.tags.className], color: base0C, fontStyle: 'italic' },
    { tag: [highlight.tags.namespace], color: base0D, fontStyle: 'italic' },
    // Operators and punctuation
    { tag: [highlight.tags.operator, highlight.tags.operatorKeyword], color: base0E },
    { tag: [highlight.tags.bracket], color: base0E },
    { tag: [highlight.tags.brace], color: base0E },
    { tag: [highlight.tags.punctuation], color: base0E },
    // Functions and parameters
    { tag: [highlight.tags.function(highlight.tags.variableName), highlight.tags.labelName], color: base0D },
    { tag: [highlight.tags.definition(highlight.tags.variableName)], color: base0D },
    // Constants and literals
    { tag: highlight.tags.number, color: base09 },
    { tag: highlight.tags.changed, color: base09 },
    { tag: highlight.tags.annotation, color: invalid, fontStyle: 'italic' },
    { tag: highlight.tags.modifier, color: base09, fontStyle: 'italic' },
    { tag: highlight.tags.self, color: base09 },
    { tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)], color: base09 },
    { tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)], color: base09 },
    // Strings and regex
    { tag: [highlight.tags.processingInstruction, highlight.tags.inserted], color: base0B },
    { tag: [highlight.tags.special(highlight.tags.string), highlight.tags.regexp], color: base0B },
    { tag: highlight.tags.string, color: base0B },
    // Punctuation and structure
    { tag: highlight.tags.definition(highlight.tags.typeName), color: base0C, fontWeight: 'bold' },
    // Comments and documentation
    { tag: highlight.tags.meta, color: base08 },
    { tag: highlight.tags.comment, fontStyle: 'italic', color: base06 },
    { tag: highlight.tags.docComment, fontStyle: 'italic', color: base06 },
    // HTML/XML elements
    { tag: [highlight.tags.tagName], color: base0A },
    { tag: [highlight.tags.attributeName], color: base0D },
    // Markdown and text formatting
    { tag: [highlight.tags.heading], fontWeight: 'bold', color: base0A },
    {
        tag: [highlight.tags.strong],
        fontWeight: 'bold',
        color: base09,
        textShadow: `0 0 2px ${base07}`,
    },
    { tag: [highlight.tags.emphasis], fontStyle: 'italic', color: base0D },
    // Links and URLs
    {
        tag: [highlight.tags.link],
        color: base0F,
        fontWeight: '500',
        textDecoration: 'underline',
        textUnderlinePosition: 'under',
    },
    {
        tag: [highlight.tags.url],
        color: base0B,
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    {
        tag: [highlight.tags.invalid],
        color: base01,
        textDecoration: 'underline wavy',
        borderBottom: `1px wavy ${invalid}`,
    },
    { tag: [highlight.tags.strikethrough], color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: highlight.tags.constant(highlight.tags.name), color: base09 },
    { tag: highlight.tags.deleted, color: invalid },
    { tag: highlight.tags.squareBracket, color: base0E },
    { tag: highlight.tags.angleBracket, color: base0E },
    // Additional specific styles
    { tag: highlight.tags.monospace, color: base01 },
    { tag: [highlight.tags.contentSeparator], color: base0D },
    { tag: highlight.tags.quote, color: base06 },
]);
/**
 * Combined Basic Dark theme extension
 */
const basicDark = [
    basicDarkTheme,
    language.syntaxHighlighting(basicDarkHighlightStyle),
];
/**
 * Basic Dark merge revert styles configuration
 */
const basicDarkMergeStyles = {
    backgroundColor: darkBackground,
    borderColor: '#3a3a3a',
    buttonColor: base05,
    buttonHoverColor: '#35373d',
};

exports.applyMergeRevertStyles = applyMergeRevertStyles;
exports.basicDark = basicDark;
exports.basicDarkMergeStyles = basicDarkMergeStyles;
