// @fsegurai/codemirror-theme-basic-light/dist/index.js
import { EditorView } from '@codemirror/view';
import { HighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { tags } from '@lezer/highlight';

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

/**
 * Enhanced color palette for Basic Light theme
 * ----------------------------------------
 * Colors are organized by function with visual color blocks for quick reference
 */
// Text colors (dark shades)
const base00 = '#1c2434', //  deep navy - primary text (darker for better contrast)
base01 = '#2d3748', //  dark slate - secondary text, cursor
base02 = '#4a5568', //  medium slate - gutter text
base03 = '#718096', //  steel blue - comments, panel text
// Background shades
base04 = '#edf2f750', // very light blue w/transparency - active line gutter
base05 = '#f7fafc', // off-white - tooltip background
base06 = '#f0f4f8', // snow-white - gutter background
// Primary accent colors (cool tones)
base07 = '#0c7792', //  teal - links, braces (more saturated)
base08 = '#0369a1', //  azure blue - numbers, constants (deeper)
base09 = '#2b6cb0', //  royal blue - variables, parameters
base0A = '#1a365d', //  deep navy - keywords, headings
// Secondary accent colors (warm and complementary)
base0B = '#c53030', //  red - square brackets (more vibrant)
base0C = '#dd6b20', //  orange - strings (warmer, more vibrant)
base0D = '#d69e2e', //  amber - class names (more saturated)
base0E = '#2f855a', //  green - operators (deeper, richer)
base0F = '#805ad5'; //  purple - tag names (more vibrant)
// UI-specific colors
const invalid = '#e53e3e', //  bright red - errors (more visible)
darkBackground = base06, // panel background
highlightBackground = '#ebf4ff40', // active line highlight (subtle blue)
background = '#ffffff', // editor background
tooltipBackground = base05, // tooltip background
selection = '#90cdf480', // selection background (clearer blue)
selectionMatch = '#63b3ed40', // selection match highlight
cursor = base01, //  cursor color
activeBracketBg = '#0c779220', // active bracket background (transparent teal)
activeBracketBorder = base09;
// Diff/merge specific colors
const addedBackground = '#e6ffec60', // Light green with transparency for insertions
removedBackground = '#ffebe9a0', // Light red with transparency for deletions
addedText = '#24783b', // Dark green for added text
removedText = '#cf222e'; // Dark red for removed text
/**
 * Enhanced editor theme styles for Basic Light
 */
const basicLightTheme = /*@__PURE__*/EditorView.theme({
    // Base editor styles
    '&': {
        color: base00,
        backgroundColor: background,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: background,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selection,
        color: base0A,
    },
    // Make sure selection appears above active line
    '.cm-selectionLayer': {
        zIndex: 100,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: '#63b3ed40',
        outline: `1px solid ${base09}`,
        color: base00,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: base00,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: '#63b3ed70',
        color: base00,
        padding: generalSearchField.padding,
        '& span': {
            color: base00,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base02,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: darkBackground,
        color: base03,
        borderRadius: '4px',
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: '1px solid #e2e8f0',
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: '1px solid #e2e8f0',
    },
    '.cm-panel button': {
        backgroundColor: darkBackground,
        color: base02,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: '#e2e8f0',
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
        zIndex: 1,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: base06,
        color: base02,
        border: generalGutter.border,
        borderRight: '1px solid #e2e8f0',
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: base04,
        color: base00,
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base02,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base00,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}30`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}30`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: '1px solid #e2e8f0',
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 2px 8px rgba(0, 0, 0, 0.1)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul': {
            backgroundColor: tooltipBackground,
            border: 'none',
        },
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: '#0c779220',
            color: base00,
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base03,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base03,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${invalid}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${base0D}`,
        },
        '&-info': {
            borderLeft: `3px solid ${base07}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${invalid}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${base0D}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: `${invalid}20`,
        outline: `1px solid ${invalid}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: selectionMatch,
        outline: `1px solid ${base09}30`,
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: 'transparent',
        color: base03,
        border: `1px dotted ${base09}50`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${background}, 0 0 0 3px ${base09}30`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: base06,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: '#cbd5e0',
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${base06}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: '#a0aec0',
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: '#718096',
    },
}, { dark: false });
/**
 * Enhanced syntax highlighting for the Basic Light theme
 */
const basicLightHighlightStyle = /*@__PURE__*/HighlightStyle.define([
    // Keywords and control flow
    { tag: tags.keyword, color: base0A, fontWeight: 'bold' },
    { tag: tags.controlKeyword, color: base0A, fontWeight: 'bold' },
    { tag: tags.moduleKeyword, color: base0A, fontWeight: 'bold' },
    // Names and variables
    { tag: [tags.name, tags.deleted, tags.character, tags.macroName], color: base09 },
    { tag: [tags.variableName], color: base09 },
    { tag: [tags.propertyName], color: base09, fontStyle: 'normal' },
    // Classes and types
    { tag: [tags.typeName], color: base0D },
    { tag: [tags.className], color: base0D, fontStyle: 'italic' },
    { tag: [tags.namespace], color: base09, fontStyle: 'italic' },
    // Operators and punctuation
    { tag: [tags.operator, tags.operatorKeyword], color: base0E },
    { tag: [tags.bracket], color: base07 },
    { tag: [tags.brace], color: base07 },
    { tag: [tags.punctuation], color: base07 },
    // Functions and parameters
    { tag: [/*@__PURE__*/tags.function(tags.variableName), tags.labelName], color: base08 },
    { tag: [/*@__PURE__*/tags.definition(tags.variableName)], color: base09 },
    // Constants and literals
    { tag: tags.number, color: base08 },
    { tag: tags.changed, color: base08 },
    { tag: tags.annotation, color: invalid, fontStyle: 'italic' },
    { tag: tags.modifier, color: base08, fontStyle: 'italic' },
    { tag: tags.self, color: base08 },
    { tag: [tags.color, /*@__PURE__*/tags.constant(tags.name), /*@__PURE__*/tags.standard(tags.name)], color: base0A },
    { tag: [tags.atom, tags.bool, /*@__PURE__*/tags.special(tags.variableName)], color: base0C },
    // Strings and regex
    { tag: [tags.processingInstruction, tags.inserted], color: base07 },
    { tag: [/*@__PURE__*/tags.special(tags.string), tags.regexp], color: base0B },
    { tag: tags.string, color: base0C },
    // Punctuation and structure
    { tag: /*@__PURE__*/tags.definition(tags.typeName), color: base0D, fontWeight: 'bold' },
    // Comments and documentation
    { tag: tags.meta, color: base08 },
    { tag: tags.comment, fontStyle: 'italic', color: base03 },
    { tag: tags.docComment, fontStyle: 'italic', color: base03 },
    // HTML/XML elements
    { tag: [tags.tagName], color: base0F },
    { tag: [tags.attributeName], color: base0D },
    // Markdown and text formatting
    { tag: [tags.heading], fontWeight: 'bold', color: base08 },
    { tag: [tags.strong], fontWeight: 'bold', color: base09 },
    { tag: [tags.emphasis], fontStyle: 'italic', color: base0A },
    // Links and URLs
    {
        tag: [tags.link],
        color: base07,
        fontWeight: '500',
        textDecoration: 'underline',
        textUnderlinePosition: 'under',
    },
    {
        tag: [tags.url],
        color: base0C,
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    {
        tag: [tags.invalid],
        color: base00,
        textDecoration: 'underline wavy',
        borderBottom: `1px wavy ${invalid}`,
    },
    { tag: [tags.strikethrough], color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: /*@__PURE__*/tags.constant(tags.name), color: base0A },
    { tag: tags.deleted, color: invalid },
    { tag: tags.squareBracket, color: base0B },
    { tag: tags.angleBracket, color: base0C },
    // Additional specific styles
    { tag: tags.heading1, fontWeight: 'bold', color: base08 },
    { tag: /*@__PURE__*/tags.special(tags.heading1), fontWeight: 'bold', color: base08 },
    {
        tag: [tags.heading2, tags.heading3, tags.heading4],
        fontWeight: 'bold',
        color: base08,
    },
    { tag: [tags.heading5, tags.heading6], color: base08 },
    { tag: tags.monospace, color: base00 },
    { tag: [tags.contentSeparator], color: base0D },
    { tag: tags.quote, color: base01 },
]);
/**
 * Combined Basic Light theme extension
 */
const basicLight = [
    basicLightTheme,
    /*@__PURE__*/syntaxHighlighting(basicLightHighlightStyle),
];
/**
 * Basic Light merge revert styles configuration
 */
const basicLightMergeStyles = {
    backgroundColor: darkBackground,
    borderColor: '#e2e8f0',
    buttonColor: base02,
    buttonHoverColor: '#e2e8f0',
};

export { applyMergeRevertStyles, basicLight, basicLightMergeStyles };
