// @fsegurai/codemirror-theme-abcdef/dist/index.cjs
'use strict';

var view = require('@codemirror/view');
var language = require('@codemirror/language');
var highlight = require('@lezer/highlight');

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

// Enhanced Abcdef theme colors with improved contrast and harmony
const base00 = '#0a0e14', // Background (slightly deeper for better contrast)
base01 = '#e0edff', // Foreground (slightly bluer tint for better readability)
base02 = '#3a4a5f', // Selection (richer blue tone)
base03 = '#242936', // Gutter background (distinguished from selection)
base04 = '#d8e1f0', // Gutter foreground (softer white)
base05 = '#4fc1ff', // Caret (bright cyan blue for visibility)
base06 = '#3a4a5f40', // Line highlight with balanced opacity
base07 = '#1a2030', // Dark background for panels
base08 = '#efbb24', // Keyword (warm gold instead of darkgoldenrod)
base09 = '#7799ff', // Atom (brighter blue)
base0A = '#8c8f93', // Comment (slightly bluer gray)
base0B = '#c792ea', // Number (softer purple)
base0C = '#ffee99', // Definition, Function (softer yellow)
base0D = '#abcdef', // Variable (keeping the theme's signature color)
base0E = '#ffcc44', // Type Name (warmer yellow)
base0F = '#99c2ff', // Tag Name (light blue)
// UI elements
invalid = '#ff5370', // Invalid color (more visible red)
darkBackground = base07, highlightBackground = base06, tooltipBackground = '#0f1521', // Darker tooltip for better contrast
cursor = base05, selection = base02, selectionForeground = '#ffffff', // Selection text color
activeBracketBg = '#3a4a5f60', // Active bracket background
activeBracketBorder = '#abcdef', // Active bracket border (signature color);
// Diff/merge specific colors
addedBackground = '#1d391d4d', // Dark green for insertions
removedBackground = '#391d1d08', // Dark red for deletions
addedText = '#7aecb3', // Bright mint green for added text
removedText = '#ff8a80'; // Soft red for removed text
/**
 * Enhanced editor theme styles for Abcdef
 */
const abcdefTheme = view.EditorView.theme({
    // Base editor styles
    '&': {
        color: base01,
        backgroundColor: base00,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: base00,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selection,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: `${base0F}40`,
        outline: `1px solid ${base0F}`,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: selectionForeground,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: '#38465b63',
        outline: `3px solid ${base0F}`,
        padding: generalSearchField.padding,
        '& span': {
            color: selectionForeground,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base04,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: darkBackground,
        color: base04,
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: '2px solid #1c2838',
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: '2px solid #1c2838',
    },
    '.cm-panel button': {
        backgroundColor: base02,
        color: base01,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: '#4a5a6f',
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: base03,
        color: base04,
        border: generalGutter.border,
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: highlightBackground,
        color: '#ffffff',
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base04,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base0D,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}40`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}40`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: '1px solid #3a4a5f',
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 6px 12px rgba(0, 0, 0, 0.3)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: '#3a4a5f',
            color: '#e0edff',
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base0D,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base0A,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${invalid}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${base08}`,
        },
        '&-info': {
            borderLeft: `3px solid ${base05}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${invalid}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${base08}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: '#3a4a5f40',
        outline: '1px solid #abcdef',
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: '#3a4a5f40',
        outline: '1px solid #abcdef',
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: '#3a4a5f40',
        color: '#abcdef',
        fontStyle: 'italic',
        border: `1px dotted ${activeBracketBorder}`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${base00}, 0 0 0 3px ${base0D}40`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: base07,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: `${base02}AA`,
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${base07}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: base02,
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: base04,
    },
}, { dark: true });
/**
 * Enhanced syntax highlighting for Abcdef theme
 */
const abcdefHighlightStyle = language.HighlightStyle.define([
    // Keywords and control flow
    { tag: highlight.tags.keyword, color: base08, fontWeight: 'bold' },
    { tag: highlight.tags.controlKeyword, color: base08, fontWeight: 'bold' },
    { tag: highlight.tags.moduleKeyword, color: base08, fontWeight: 'bold' },
    // Names and variables
    { tag: [highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName], color: base09 },
    { tag: [highlight.tags.variableName], color: base09 },
    { tag: [highlight.tags.propertyName], color: base0D, fontStyle: 'normal' },
    // Classes and types
    { tag: highlight.tags.typeName, color: base0E },
    { tag: highlight.tags.className, color: base0D, fontStyle: 'italic' },
    { tag: highlight.tags.namespace, color: '#78a0d3', opacity: 0.8 },
    // Operators and punctuation - clearer blues
    { tag: [highlight.tags.operator, highlight.tags.operatorKeyword], color: '#ff9cac' },
    { tag: [highlight.tags.bracket], color: '#d0d6e0' },
    { tag: [highlight.tags.brace], color: '#d0d6e0' },
    { tag: [highlight.tags.punctuation], color: '#d0d6e0' },
    // Functions and parameters
    { tag: highlight.tags.function(highlight.tags.variableName), color: base0C },
    { tag: highlight.tags.definition(highlight.tags.variableName), color: base0C },
    // Constants and literals
    { tag: highlight.tags.number, color: base0B },
    { tag: highlight.tags.changed, color: '#ff9cac' },
    { tag: highlight.tags.annotation, color: '#ffad5c', fontStyle: 'italic' },
    { tag: highlight.tags.modifier, color: '#ffad5c', fontStyle: 'italic' },
    { tag: highlight.tags.self, color: base0B },
    { tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)], color: base0B },
    { tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)], color: base0B },
    // Strings and regex
    { tag: [highlight.tags.processingInstruction, highlight.tags.inserted], color: '#7aecb3' },
    { tag: [highlight.tags.special(highlight.tags.string), highlight.tags.regexp], color: '#6acdbe' /* Turquoise */ },
    // Punctuation and structure
    { tag: highlight.tags.definition(highlight.tags.typeName), color: base0E, fontWeight: 'bold' },
    { tag: [highlight.tags.operator, highlight.tags.operatorKeyword], color: '#ff9cac' },
    { tag: highlight.tags.bracket, color: '#d0d6e0' },
    { tag: [highlight.tags.brace], color: '#d0d6e0' },
    { tag: highlight.tags.punctuation, color: '#d0d6e0' },
    // Comments and documentation
    { tag: highlight.tags.meta, color: '#78a0d3' },
    { tag: highlight.tags.comment, fontStyle: 'italic', color: base0A },
    { tag: highlight.tags.docComment, fontStyle: 'italic', color: base0A },
    // HTML/XML elements
    { tag: highlight.tags.tagName, color: base0F },
    { tag: highlight.tags.attributeName, color: '#ffad5c' },
    // Markdown and text formatting
    { tag: highlight.tags.heading, color: base0E, fontWeight: 'bold' },
    { tag: [highlight.tags.strong], fontWeight: 'bold' },
    { tag: [highlight.tags.emphasis], fontStyle: 'italic' },
    // Links and URLs
    { tag: highlight.tags.link, color: '#ffcc44', fontWeight: '500' },
    {
        tag: highlight.tags.url,
        color: '#66c2ff',
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    { tag: highlight.tags.invalid, color: invalid, textDecoration: 'underline wavy' },
    { tag: highlight.tags.strikethrough, color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: highlight.tags.constant(highlight.tags.name), color: base0B },
    { tag: highlight.tags.controlKeyword, color: base08, fontWeight: 'bold' },
    { tag: highlight.tags.deleted, color: invalid },
    { tag: highlight.tags.labelName, color: '#ffad5c' },
    { tag: highlight.tags.string, color: '#7aecb3' /* New mint green for strings */ },
]);
/**
 * Combined Abcdef theme extension
 */
const abcdef = [
    abcdefTheme,
    language.syntaxHighlighting(abcdefHighlightStyle),
];
/**
 * Abcdef merge revert styles configuration
 */
const abcdefMergeStyles = {
    backgroundColor: base03, // Using theme-specific color variables
    borderColor: base02,
    buttonColor: selectionForeground,
    buttonHoverColor: base02,
};

exports.abcdef = abcdef;
exports.abcdefMergeStyles = abcdefMergeStyles;
exports.applyMergeRevertStyles = applyMergeRevertStyles;
