// @fsegurai/codemirror-theme-gruvbox-light/dist/index.js
import { EditorView } from '@codemirror/view';
import { HighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { tags } from '@lezer/highlight';

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

/**
 * Enhanced Gruvbox Light theme color definitions
 * --------------------------------------------
 * Colors organized by function with visual color blocks
 */
// Gruvbox base colors
const base00 = '#3c3836', // Main foreground (text)
base01 = '#504945', // Secondary foreground
base02 = '#665c54', // Tertiary foreground
base03 = '#7c6f64', // Quaternary foreground
base04 = '#928374', // Comments, invisibles, line highlighting
// Light/background shades
base05 = '#fbf1c7', // Main background
base06 = '#ebdbb2', // Secondary background
base07 = '#d5c4a1', // Tertiary background (not defined above)
base08 = '#bdae93', // Quaternary background
base09 = '#a89984', // Quinary background (not defined above)
// Accent colors
base0A = '#9d0006', // Keywords, storage, operator
base0B = '#79740e', // Strings, tag attributes
base0C = '#b57614', // Functions, tag names
base0D = '#076678', // Variables
base0E = '#8f3f71', // Numbers, special constants
base0F = '#427b58', // Types
base10 = '#af3a03'; // Cursor, constants
// UI specific colors
const invalid = base0A, darkBackground = base06, highlightBackground = '#ffc42e25', // Line highlight with transparency
background = base05, tooltipBackground = base06, selection = darkBackground, selectionMatch = '#ffc42e40', // Selection match background
cursor = base10, // Cursor color
activeBracketBg = '#d5c4a180', // Active bracket background
activeBracketBorder = base10, // Active bracket border
diagnosticWarning = base0C, // Warning color
linkColor = base0D, // Link color
visitedLinkColor = base0E; // Visited link color
// Diff/merge specific colors
const addedBackground = '#d8e5bc80', // Light base0B with transparency for insertions
removedBackground = '#f7cfcf80', // Light base0A with transparency for deletions
addedText = '#79740e', // Gruvbox base0B for added text
removedText = '#9d0006'; // Gruvbox base0A for removed text
/**
 * Enhanced editor theme styles for Gruvbox Light
 */
const gruvboxLightTheme = /*@__PURE__*/EditorView.theme({
    // Base editor styles
    '&': {
        color: base00,
        backgroundColor: background,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: base05,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selection,
        color: base00,
    },
    // Make sure selection appears above active line
    '.cm-selectionLayer': {
        zIndex: 100,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: '#ffc42e80',
        outline: `1px solid ${base10}`,
        color: base00,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: base00,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: base10,
        color: base05,
        padding: generalSearchField.padding,
        '& span': {
            color: base05,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base00,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: base06,
        color: base01,
        borderRadius: '0 0 4px 4px',
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: `1px solid ${base08}`,
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: `1px solid ${base08}`,
    },
    '.cm-panel button': {
        backgroundColor: base05,
        color: base00,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: base07,
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
        zIndex: 1,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: base06,
        color: base03,
        border: generalGutter.border,
        borderRight: `1px solid ${base08}`,
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: base07,
        color: base00,
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base03,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base00,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}40`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}40`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: `1px solid ${base08}`,
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 1px 5px rgba(0, 0, 0, 0.1)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul': {
            backgroundColor: tooltipBackground,
            border: 'none',
        },
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: base10 + '30',
            color: base00,
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li:hover': {
            backgroundColor: base10 + '15',
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base03,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base03,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${invalid}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${diagnosticWarning}`,
        },
        '&-info': {
            borderLeft: `3px solid ${linkColor}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${invalid}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${diagnosticWarning}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: '#fb492740',
        outline: `1px solid ${invalid}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: selectionMatch,
        outline: `1px solid ${base08}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: 'transparent',
        color: base02,
        border: `1px dotted ${base08}`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${base05}, 0 0 0 3px ${base10}40`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: base05,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: base08,
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${base05}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: base09,
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: base04,
    },
}, { dark: false });
/**
 * Enhanced syntax highlighting for Gruvbox Light theme
 */
const gruvboxLightHighlightStyle = /*@__PURE__*/HighlightStyle.define([
    // Keywords and control flow
    { tag: tags.keyword, color: base0A, fontWeight: 'bold' },
    { tag: tags.controlKeyword, color: base0A, fontWeight: 'bold' },
    { tag: tags.moduleKeyword, color: base0A, fontWeight: 'bold' },
    // Names and variables
    { tag: [tags.name, tags.deleted, tags.character, tags.macroName], color: base0D },
    { tag: [tags.variableName], color: base0D },
    { tag: [tags.propertyName], color: base0F, fontStyle: 'normal' },
    // Classes and types
    { tag: [tags.typeName], color: base0F },
    { tag: [tags.className], color: base0C, fontStyle: 'italic' },
    { tag: [tags.namespace], color: base0D, fontStyle: 'italic' },
    // Operators and punctuation
    { tag: [tags.operator, tags.operatorKeyword], color: base00 },
    { tag: [tags.bracket], color: base04 },
    { tag: [tags.brace], color: base04 },
    { tag: [tags.punctuation], color: base04 },
    // Functions and parameters
    { tag: [/*@__PURE__*/tags.function(tags.variableName), tags.labelName], color: base0C },
    { tag: [/*@__PURE__*/tags.definition(tags.variableName)], color: base0D },
    // Constants and literals
    { tag: tags.number, color: base0E },
    { tag: tags.changed, color: base0E },
    { tag: tags.annotation, color: invalid, fontStyle: 'italic' },
    { tag: tags.modifier, color: base0E, fontStyle: 'italic' },
    { tag: tags.self, color: base0E },
    { tag: [tags.color, /*@__PURE__*/tags.constant(tags.name), /*@__PURE__*/tags.standard(tags.name)], color: base10 },
    { tag: [tags.atom, tags.bool, /*@__PURE__*/tags.special(tags.variableName)], color: base10 },
    // Strings and regex
    { tag: [tags.processingInstruction, tags.inserted], color: base0B },
    { tag: [/*@__PURE__*/tags.special(tags.string), tags.regexp], color: base0B },
    { tag: tags.string, color: base0B },
    // Punctuation and structure
    { tag: /*@__PURE__*/tags.definition(tags.typeName), color: base0F, fontWeight: 'bold' },
    // Comments and documentation
    { tag: tags.meta, color: base04 },
    { tag: tags.comment, fontStyle: 'italic', color: base04 },
    { tag: tags.docComment, fontStyle: 'italic', color: base04 },
    // HTML/XML elements
    { tag: [tags.tagName], color: base0A },
    { tag: [tags.attributeName], color: base0C },
    // Markdown and text formatting
    { tag: [tags.heading], fontWeight: 'bold', color: base0C },
    { tag: [tags.strong], fontWeight: 'bold', color: base0C },
    { tag: [tags.emphasis], fontStyle: 'italic', color: base0B },
    // Links and URLs
    {
        tag: [tags.link],
        color: visitedLinkColor,
        fontWeight: '500',
        textDecoration: 'underline',
        textUnderlinePosition: 'under',
    },
    {
        tag: [tags.url],
        color: linkColor,
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    {
        tag: [tags.invalid],
        color: base00,
        textDecoration: 'underline wavy',
        borderBottom: `1px wavy ${invalid}`,
    },
    { tag: [tags.strikethrough], color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: /*@__PURE__*/tags.constant(tags.name), color: base10 },
    { tag: tags.deleted, color: invalid },
    { tag: tags.squareBracket, color: base04 },
    { tag: tags.angleBracket, color: base04 },
    // Additional specific styles
    { tag: tags.monospace, color: base00 },
    { tag: [tags.contentSeparator], color: base0D },
    { tag: tags.quote, color: base04 },
]);
/**
 * Combined Gruvbox Light theme extension
 */
const gruvboxLight = [
    gruvboxLightTheme,
    /*@__PURE__*/syntaxHighlighting(gruvboxLightHighlightStyle),
];
/**
 * Gruvbox Light merge revert styles configuration
 */
const gruvboxLightMergeStyles = {
    backgroundColor: tooltipBackground,
    borderColor: base08,
    buttonColor: base00,
    buttonHoverColor: base07,
};

export { applyMergeRevertStyles, gruvboxLight, gruvboxLightMergeStyles };
