// @fsegurai/codemirror-theme-solarized-light/dist/index.cjs
'use strict';

var view = require('@codemirror/view');
var language = require('@codemirror/language');
var highlight = require('@lezer/highlight');

// Helper module for styling options
const generalContent = {
    fontSize: '14px',
    fontFamily: 'JetBrains Mono, Consolas, monospace',
    lineHeight: '1.6',
};
const generalCursor = {
    borderLeftWidth: '2px',
};
const generalDiff = {
    insertedTextDecoration: 'none',
    deletedTextDecoration: 'line-through',
    insertedLinePadding: '1px 3px',
    borderRadious: '3px'};
const generalGutter = {
    border: 'none',
    paddingRight: '8px',
    fontSize: '0.9em',
    fontWeight: '500',
};
const generalPanel = {
    border: 'none',
    borderRadius: '4px',
    padding: '2px 10px',
};
const generalLine = {
    borderRadius: '2px',
};
const generalMatching = {
    borderRadius: '2px',
};
const generalPlaceholder = {
    borderRadius: '4px',
    padding: '0 5px',
    margin: '0 2px',
};
const generalScroller = {
    width: '12px',
    height: '12px',
    borderRadius: '6px',
};
const generalSearchField = {
    borderRadius: '4px',
    padding: '2px 6px',
};
const generalTooltip = {
    borderRadius: '4px',
    borderRadiusSelected: '3px',
    lineHeight: '1.3',
    padding: '4px 8px',
    paddingRight: '8px',
};
/**
 * Function to apply merge revert styles for a theme
 * @param styles Styles for the merge revert buttons
 * @param styles.backgroundColor Background color of the revert area
 * @param styles.borderColor Border color of the revert area
 * @param styles.buttonColor Color of the revert buttons
 * @param styles.buttonHoverColor Hover color of the revert buttons
 */
function applyMergeRevertStyles(styles) {
    // Create a stylesheet
    const styleEl = document.createElement('style');
    styleEl.id = 'cm-merge-revert-styles';
    // Define CSS with the theme-specific values
    styleEl.textContent = `
    .cm-merge-revert {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: flex-start;
      padding: 4px;
      background-color: ${styles.backgroundColor};
      border-left: 1px solid ${styles.borderColor};
      border-right: 1px solid ${styles.borderColor};
      width: 32px;
    }
    
    .cm-merge-revert button {
      width: 100%;
      height: auto;
      background-color: transparent;
      border: none;
      color: ${styles.buttonColor};
      cursor: pointer;
      margin: 0 auto;
      font-size: 20px;
    }
    
    .cm-merge-revert button:hover {
      background-color: ${styles.buttonHoverColor};
    }
  `;
    // Remove any existing merge styles
    const existingStyle = document.getElementById('cm-merge-revert-styles');
    if (existingStyle)
        existingStyle.remove();
    // Add the new styles
    document.head.appendChild(styleEl);
}

/**
 * Enhanced Solarized Light theme color definitions
 * ---------------------------------------------
 * Colors organized by function with visual color blocks
 */
// Base colors from Solarized palette
const base00 = '#657b83', // Body text/default text color
base01 = '#586e75', // Optional emphasized content
base02 = '#073642', // Background highlights
base03 = '#002b36', // Comments, invisible, line highlighting
base04 = '#dfd9c8', // Dark background tint
base05 = '#93a1a1', // Default foreground/UI text color
base06 = '#cceeff7a', // Light background tint (for selection)
base07 = '#fdf6e3', // Background - light base
base08 = '#eee8d5', // Background tint - light secondary
// Accent colors from Solarized palette
base09 = '#dc322f', // Red
base0A = '#cb4b16', // Orange
base0B = '#b58900', // Yellow
base0C = '#859900', // Green
base0D = '#2aa198', // Cyan
base0E = '#268bd2', // Blue
base0F = '#6c71c4', // Violet
base10 = '#d33682'; // Magenta
// UI specific colors
const invalid = '#d30102', // Bright red for errors
darkBackground = base04, // Darker background for panels
highlightBackground = base06, // Active line highlight
background = base07, // Main editor background
tooltipBackground = '#f0e9d7', // Tooltip background
selection = '#ffd07a', // Selection background
selectionMatch = '#e1dbca90', // Selection match with opacity
cursor = base01, // Cursor color
activeBracketBg = '#93a1a140', // Active bracket background with opacity
activeBracketBorder = base0E, // Active bracket border - blue
diagnosticWarning = base0B, // Warning color - yellow
linkColor = base0E; // Link color - blue
// Diff/merge specific colors
const addedBackground = '#85990020', // Solarized green with transparency for insertions
removedBackground = '#dc322f20', // Solarized red with transparency for deletions
addedText = '#859900', // Solarized green for added text
removedText = '#dc322f'; // Solarized red for removed text
/**
 * Enhanced editor theme styles for Solarized Light
 */
const solarizedLightTheme = view.EditorView.theme({
    // Base editor styles
    '&': {
        color: base00,
        backgroundColor: background,
        fontSize: generalContent.fontSize,
        fontFamily: generalContent.fontFamily,
    },
    // Content and cursor
    '.cm-content': {
        caretColor: cursor,
        lineHeight: generalContent.lineHeight,
    },
    '.cm-cursor, .cm-dropCursor': {
        borderLeftColor: cursor,
        borderLeftWidth: generalCursor.borderLeftWidth,
    },
    '.cm-fat-cursor': {
        backgroundColor: `${cursor}99`,
        color: background,
    },
    // Selection
    '&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection': {
        backgroundColor: selection,
    },
    // Make sure selection appears above active line
    '.cm-selectionLayer': {
        zIndex: 100,
    },
    // Search functionality
    '.cm-searchMatch': {
        backgroundColor: '#93a1a180',
        outline: `1px solid ${base0E}`,
        color: base01,
        borderRadius: generalSearchField.borderRadius,
        '& span': {
            color: base01,
        },
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
        backgroundColor: base0E,
        color: base08,
        padding: generalSearchField.padding,
        '& span': {
            color: base08,
        },
    },
    '.cm-search.cm-panel.cm-textfield': {
        color: base00,
        borderRadius: generalSearchField.borderRadius,
        padding: generalSearchField.padding,
    },
    // Panels
    '.cm-panels': {
        backgroundColor: darkBackground,
        color: base00,
        borderRadius: '4px',
    },
    '.cm-panels.cm-panels-top': {
        borderBottom: `1px solid ${base04}`,
    },
    '.cm-panels.cm-panels-bottom': {
        borderTop: `1px solid ${base04}`,
    },
    '.cm-panel button': {
        backgroundColor: tooltipBackground,
        color: base00,
        border: generalPanel.border,
        borderRadius: generalPanel.borderRadius,
        padding: generalPanel.padding,
    },
    '.cm-panel button:hover': {
        backgroundColor: highlightBackground,
    },
    // Line highlighting
    '.cm-activeLine': {
        backgroundColor: highlightBackground,
        borderRadius: generalLine.borderRadius,
        zIndex: 1,
    },
    // Gutters
    '.cm-gutters': {
        backgroundColor: darkBackground,
        color: base01,
        border: generalGutter.border,
        borderRight: `1px solid ${base04}`,
        paddingRight: generalGutter.paddingRight,
    },
    '.cm-activeLineGutter': {
        backgroundColor: highlightBackground,
        color: base00,
        fontWeight: generalGutter.fontWeight,
    },
    '.cm-lineNumbers': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter': {
        fontSize: generalGutter.fontSize,
    },
    '.cm-foldGutter .cm-gutterElement': {
        color: base01,
        cursor: 'pointer',
    },
    '.cm-foldGutter .cm-gutterElement:hover': {
        color: base00,
    },
    // Diff/Merge View Styles
    // Inserted/Added Content
    '.cm-insertedLine': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: addedBackground,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'ins.cm-insertedLine, ins.cm-insertedLine:not(:has(.cm-changedText))': {
        textDecoration: generalDiff.insertedTextDecoration,
        backgroundColor: `${addedBackground} !important`,
        color: addedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${addedText}40`,
    },
    'ins.cm-insertedLine .cm-changedText': {
        background: 'transparent !important',
    },
    // Deleted/Removed Content
    '.cm-deletedLine': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: removedBackground,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
    },
    'del.cm-deletedLine, del, del:not(:has(.cm-deletedText))': {
        textDecoration: generalDiff.deletedTextDecoration,
        backgroundColor: `${removedBackground} !important`,
        color: removedText,
        padding: generalDiff.insertedLinePadding,
        borderRadius: generalDiff.borderRadious,
        border: `1px solid ${removedText}40`,
    },
    'del .cm-deletedText, del .cm-changedText': {
        background: 'transparent !important',
    },
    // Tooltips and autocomplete
    '.cm-tooltip': {
        backgroundColor: tooltipBackground,
        border: `1px solid ${base01}`,
        borderRadius: generalTooltip.borderRadius,
        padding: generalTooltip.padding,
        boxShadow: '0 2px 8px rgba(0, 0, 0, 0.15)',
    },
    '.cm-tooltip-autocomplete': {
        '& > ul': {
            backgroundColor: tooltipBackground,
            border: 'none',
        },
        '& > ul > li': {
            padding: generalTooltip.padding,
            lineHeight: generalTooltip.lineHeight,
        },
        '& > ul > li[aria-selected]': {
            backgroundColor: selection,
            color: base01,
            borderRadius: generalTooltip.borderRadiusSelected,
        },
        '& > ul > li > span.cm-completionIcon': {
            color: base01,
            paddingRight: generalTooltip.paddingRight,
        },
        '& > ul > li > span.cm-completionDetail': {
            color: base01,
            fontStyle: 'italic',
        },
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
        borderTopColor: tooltipBackground,
        borderBottomColor: tooltipBackground,
    },
    // Diagnostics styling
    '.cm-diagnostic': {
        '&-error': {
            borderLeft: `3px solid ${base09}`,
        },
        '&-warning': {
            borderLeft: `3px solid ${diagnosticWarning}`,
        },
        '&-info': {
            borderLeft: `3px solid ${linkColor}`,
        },
    },
    '.cm-lintPoint-error': {
        borderBottom: `2px wavy ${base09}`,
    },
    '.cm-lintPoint-warning': {
        borderBottom: `2px wavy ${diagnosticWarning}`,
    },
    // Matching brackets
    '.cm-matchingBracket': {
        backgroundColor: activeBracketBg,
        outline: `1px solid ${activeBracketBorder}`,
        borderRadius: generalMatching.borderRadius,
    },
    '.cm-nonmatchingBracket': {
        backgroundColor: `${base09}40`,
        outline: `1px solid ${invalid}`,
        borderRadius: generalMatching.borderRadius,
    },
    // Selection matches
    '.cm-selectionMatch': {
        backgroundColor: selectionMatch,
        outline: `1px solid ${base05}50`,
        borderRadius: generalMatching.borderRadius,
    },
    // Fold placeholder
    '.cm-foldPlaceholder': {
        backgroundColor: tooltipBackground,
        color: base03,
        border: `1px dotted ${base05}70`,
        borderRadius: generalPlaceholder.borderRadius,
        padding: generalPlaceholder.padding,
        margin: generalPlaceholder.margin,
    },
    // Focus outline
    '&.cm-focused': {
        outline: 'none',
        boxShadow: `0 0 0 2px ${background}, 0 0 0 3px ${base0E}40`,
    },
    // Scrollbars
    '& .cm-scroller::-webkit-scrollbar': {
        width: generalScroller.width,
        height: generalScroller.height,
    },
    '& .cm-scroller::-webkit-scrollbar-track': {
        background: darkBackground,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb': {
        backgroundColor: base04,
        borderRadius: generalScroller.borderRadius,
        border: `3px solid ${darkBackground}`,
    },
    '& .cm-scroller::-webkit-scrollbar-thumb:hover': {
        backgroundColor: base05,
    },
    // Ghost text
    '.cm-ghostText': {
        opacity: '0.5',
        color: base01,
    },
}, { dark: false });
/**
 * Enhanced syntax highlighting for the Solarized Light theme
 */
const solarizedLightHighlightStyle = language.HighlightStyle.define([
    // Keywords and control flow
    { tag: highlight.tags.keyword, color: base0C, fontWeight: 'bold' },
    { tag: highlight.tags.controlKeyword, color: base0C, fontWeight: 'bold' },
    { tag: highlight.tags.moduleKeyword, color: base0C, fontWeight: 'bold' },
    // Names and variables
    { tag: [highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName], color: base0D },
    { tag: [highlight.tags.variableName], color: base00 },
    { tag: [highlight.tags.propertyName], color: base0D, fontStyle: 'normal' },
    // Classes and types
    { tag: [highlight.tags.typeName], color: base0A },
    { tag: [highlight.tags.className], color: base0A, fontStyle: 'italic' },
    { tag: [highlight.tags.namespace], color: base10, fontStyle: 'italic' },
    // Operators and punctuation
    { tag: [highlight.tags.operator, highlight.tags.operatorKeyword], color: base0F },
    { tag: [highlight.tags.bracket], color: base10 },
    { tag: [highlight.tags.brace], color: base10 },
    { tag: [highlight.tags.punctuation], color: base01 },
    // Functions and parameters
    { tag: [highlight.tags.function(highlight.tags.variableName)], color: base0E },
    { tag: [highlight.tags.labelName], color: base10 },
    { tag: [highlight.tags.definition(highlight.tags.function(highlight.tags.variableName))], color: base0E },
    { tag: [highlight.tags.definition(highlight.tags.variableName)], color: base0D },
    // Constants and literals
    { tag: highlight.tags.number, color: base10 },
    { tag: highlight.tags.changed, color: base10 },
    { tag: highlight.tags.annotation, color: invalid, fontStyle: 'italic' },
    { tag: highlight.tags.modifier, color: base10, fontStyle: 'italic' },
    { tag: highlight.tags.self, color: base10 },
    {
        tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)],
        color: base0B,
    },
    { tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)], color: base10 },
    // Strings and regex
    { tag: [highlight.tags.processingInstruction, highlight.tags.inserted], color: base0C },
    { tag: [highlight.tags.special(highlight.tags.string), highlight.tags.regexp], color: invalid },
    { tag: highlight.tags.string, color: base0B },
    // Punctuation and structure
    { tag: highlight.tags.definition(highlight.tags.typeName), color: base0A, fontWeight: 'bold' },
    { tag: [highlight.tags.definition(highlight.tags.name), highlight.tags.separator], color: base0D },
    // Comments and documentation
    { tag: highlight.tags.meta, color: base09 },
    { tag: highlight.tags.comment, fontStyle: 'italic', color: base01 },
    { tag: highlight.tags.docComment, fontStyle: 'italic', color: base01 },
    // HTML/XML elements
    { tag: [highlight.tags.tagName], color: base0E },
    { tag: [highlight.tags.attributeName], color: base00 },
    // Markdown and text formatting
    { tag: [highlight.tags.heading], fontWeight: 'bold', color: base0B },
    { tag: highlight.tags.heading1, color: base03 },
    { tag: highlight.tags.heading2, color: base02 },
    { tag: highlight.tags.heading3, color: base02 },
    { tag: highlight.tags.heading4, color: base02 },
    { tag: highlight.tags.heading5, color: base02 },
    { tag: highlight.tags.heading6, color: base02 },
    { tag: [highlight.tags.strong], fontWeight: 'bold', color: base02 },
    { tag: [highlight.tags.emphasis], fontStyle: 'italic', color: base0C },
    // Links and URLs
    {
        tag: [highlight.tags.link],
        color: base0D,
        fontWeight: '500',
        textDecoration: 'underline',
        textUnderlinePosition: 'under',
    },
    {
        tag: [highlight.tags.url],
        color: base0B,
        textDecoration: 'underline',
        textUnderlineOffset: '2px',
    },
    // Special states
    {
        tag: [highlight.tags.invalid],
        color: base01,
        textDecoration: 'underline wavy',
        borderBottom: `1px dotted ${base09}`,
    },
    { tag: [highlight.tags.strikethrough], color: invalid, textDecoration: 'line-through' },
    // Enhanced syntax highlighting
    { tag: highlight.tags.constant(highlight.tags.name), color: base0B },
    { tag: highlight.tags.deleted, color: base09 },
    { tag: highlight.tags.squareBracket, color: base09 },
    { tag: highlight.tags.angleBracket, color: base01 },
    // Additional specific styles
    { tag: highlight.tags.monospace, color: base00 },
    { tag: [highlight.tags.contentSeparator], color: base0B },
    { tag: highlight.tags.quote, color: base0C },
]);
/**
 * Combined Solarized Light theme extension
 */
const solarizedLight = [
    solarizedLightTheme,
    language.syntaxHighlighting(solarizedLightHighlightStyle),
];
/**
 * Solarized Light merge revert styles configuration
 */
const solarizedLightMergeStyles = {
    backgroundColor: base04,
    borderColor: base02,
    buttonColor: base0D,
    buttonHoverColor: `${base05}40`,
};

exports.applyMergeRevertStyles = applyMergeRevertStyles;
exports.solarizedLight = solarizedLight;
exports.solarizedLightMergeStyles = solarizedLightMergeStyles;
